File: WeatheriOS/Weather/Services/GeocodingService.swift
//
//  GeocodingService.swift
//  Weather
//
//  Created by Rick Pasetto on 12/21/25.
//

import Foundation
import CoreLocation
import Contacts

class GeocodingService {
    private let geocoder = CLGeocoder()
    
    func geocode(cityName: String) async throws -> City {
        let placemarks = try await geocoder.geocodeAddressString(cityName)
        
        guard let placemark = placemarks.first,
              let location = placemark.location else {
            throw GeocodingError.locationNotFound
        }
        
        let name = placemark.locality ?? placemark.administrativeArea ?? cityName
        let state = extractStateAbbreviation(from: placemark)
        
        return City(
            name: name,
            state: state,
            latitude: location.coordinate.latitude,
            longitude: location.coordinate.longitude
        )
    }
    
    private func extractStateAbbreviation(from placemark: CLPlacemark) -> String? {
        if let postalAddress = placemark.postalAddress {
            return postalAddress.state
        }
        
        if let administrativeArea = placemark.administrativeArea {
            return StateAbbreviationMapper.abbreviation(for: administrativeArea)
        }
        
        return nil
    }
}

struct StateAbbreviationMapper {
    private static let stateMap: [String: String] = [
        "Alabama": "AL", "Alaska": "AK", "Arizona": "AZ", "Arkansas": "AR",
        "California": "CA", "Colorado": "CO", "Connecticut": "CT", "Delaware": "DE",
        "Florida": "FL", "Georgia": "GA", "Hawaii": "HI", "Idaho": "ID",
        "Illinois": "IL", "Indiana": "IN", "Iowa": "IA", "Kansas": "KS",
        "Kentucky": "KY", "Louisiana": "LA", "Maine": "ME", "Maryland": "MD",
        "Massachusetts": "MA", "Michigan": "MI", "Minnesota": "MN", "Mississippi": "MS",
        "Missouri": "MO", "Montana": "MT", "Nebraska": "NE", "Nevada": "NV",
        "New Hampshire": "NH", "New Jersey": "NJ", "New Mexico": "NM", "New York": "NY",
        "North Carolina": "NC", "North Dakota": "ND", "Ohio": "OH", "Oklahoma": "OK",
        "Oregon": "OR", "Pennsylvania": "PA", "Rhode Island": "RI", "South Carolina": "SC",
        "South Dakota": "SD", "Tennessee": "TN", "Texas": "TX", "Utah": "UT",
        "Vermont": "VT", "Virginia": "VA", "Washington": "WA", "West Virginia": "WV",
        "Wisconsin": "WI", "Wyoming": "WY", "District of Columbia": "DC"
    ]
    
    static func abbreviation(for stateName: String) -> String? {
        return stateMap[stateName] ?? stateName
    }
}

enum GeocodingError: LocalizedError {
    case locationNotFound
    
    var errorDescription: String? {
        switch self {
        case .locationNotFound:
            return "Could not find location for the city name"
        }
    }
}

File: WeatherAndroid/app/src/main/java/com/weatherdemo/services/GeocodingService.kt
package com.weatherdemo.services

import android.location.Geocoder
import com.weatherdemo.models.City
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class GeocodingService(private val geocoder: Geocoder) {
    
    suspend fun geocode(cityName: String): City = withContext(Dispatchers.IO) {
        @Suppress("DEPRECATION")
        val addresses = geocoder.getFromLocationName(cityName, 1)
        
        if (addresses.isNullOrEmpty()) {
            throw GeocodingException("Location not found")
        }
        
        val address = addresses.first() ?: throw GeocodingException("Location not found")
        
        val name = address.locality ?: address.adminArea ?: cityName
        val state = extractStateAbbreviation(address)
        val latitude = address.latitude
        val longitude = address.longitude
        
        City(
            name = name,
            state = state,
            latitude = latitude,
            longitude = longitude
        )
    }
    
    private fun extractStateAbbreviation(address: android.location.Address): String? {
        val state = address.adminArea ?: return null
        return StateAbbreviationMapper.abbreviation(state)
    }
}

object StateAbbreviationMapper {
    private val stateMap = mapOf(
        "Alabama" to "AL", "Alaska" to "AK", "Arizona" to "AZ", "Arkansas" to "AR",
        "California" to "CA", "Colorado" to "CO", "Connecticut" to "CT", "Delaware" to "DE",
        "Florida" to "FL", "Georgia" to "GA", "Hawaii" to "HI", "Idaho" to "ID",
        "Illinois" to "IL", "Indiana" to "IN", "Iowa" to "IA", "Kansas" to "KS",
        "Kentucky" to "KY", "Louisiana" to "LA", "Maine" to "ME", "Maryland" to "MD",
        "Massachusetts" to "MA", "Michigan" to "MI", "Minnesota" to "MN", "Mississippi" to "MS",
        "Missouri" to "MO", "Montana" to "MT", "Nebraska" to "NE", "Nevada" to "NV",
        "New Hampshire" to "NH", "New Jersey" to "NJ", "New Mexico" to "NM", "New York" to "NY",
        "North Carolina" to "NC", "North Dakota" to "ND", "Ohio" to "OH", "Oklahoma" to "OK",
        "Oregon" to "OR", "Pennsylvania" to "PA", "Rhode Island" to "RI", "South Carolina" to "SC",
        "South Dakota" to "SD", "Tennessee" to "TN", "Texas" to "TX", "Utah" to "UT",
        "Vermont" to "VT", "Virginia" to "VA", "Washington" to "WA", "West Virginia" to "WV",
        "Wisconsin" to "WI", "Wyoming" to "WY", "District of Columbia" to "DC"
    )
    
    fun abbreviation(stateName: String): String? {
        return stateMap[stateName] ?: stateName
    }
}

class GeocodingException(message: String) : Exception(message)

